File: scalechain-proto-codec/src/test/scala/io/scalechain/blockchain/proto/codec/EnvelopeSuite.scala
package io.scalechain.blockchain.proto.codec
import io.scalechain.blockchain.proto._
import io.scalechain.util.HexUtil._

import org.scalatest.{FlatSpec, ShouldMatchers}
import scodec.Codec
import scodec.bits.BitVector


/**
  * Created by kangmo on 1/29/16.
  */
trait PayloadTestSuite[T] extends FlatSpec with ShouldMatchers with CodecTestUtil {
  implicit val codec : Codec[T]

  val message : T
  val payload : Array[Byte]

  def payloadBits() = BitVector.view(payload)

  "payload" should "be correctly encoded" in {
    if (message != null) {
      val encodedBits = encode(message)

      println("Encoded bits : " + encodedBits.bytes.toString)

      encodedBits shouldBe payloadBits
    }
  }

  "payload" should "be correctly decoded" in {
    val actual = decodeFully(payloadBits)

    println("actual message = " + actual)
    println("expected message = " + message)

    if (message != null) {
      decodeFully(payloadBits) shouldBe message
    }
  }

  "payload" should "be correctly decoded and encoded" in {
    val obj = decodeFully(payloadBits)
    encode(obj) shouldBe payloadBits
  }
}


/**
  * Created by kangmo on 1/29/16.
  */
trait EnvelopeTestSuite[T] extends PayloadTestSuite[T] {
  val envelopeCodec = BitcoinMessageEnvelope.codec

  val envelope : BitcoinMessageEnvelope
  val envelopeHeader : Array[Byte]

  def envelopeBits() = BitVector.view(envelopeHeader) ++ payloadBits

  "envelope" should "be correctly encoded" in {
    envelopeCodec.encode(envelope).require shouldBe envelopeBits
  }

  "envelope" should "be correctly decoded" in {
    val actual = envelopeCodec.decode(envelopeBits).require.value
    envelopeCodec.decode(envelopeBits).require.value shouldBe envelope
  }
}

File: scalechain-proto-codec/src/test/kotlin/io/scalechain/blockchain/proto/codec/EnvelopeSuite.kt
package io.scalechain.blockchain.proto.codec
import io.scalechain.blockchain.proto.*
import io.scalechain.util.HexUtil
import io.scalechain.util.HexUtil.bytes
import io.scalechain.util.HexUtil.hex
import io.kotlintest.*
import io.kotlintest.matchers.Matchers
import io.kotlintest.properties.Table2
import io.kotlintest.specs.FlatSpec
import io.netty.buffer.ByteBuf
import io.netty.buffer.ByteBufUtil
import io.scalechain.util.toByteArray

/**
  * Created by kangmo on 1/29/16.
  */
abstract class PayloadTestSuite<T> : FlatSpec(), Matchers, CodecTestUtil {
  abstract val codec : Codec<T>
  abstract val message : T?
  abstract val payload : ByteArray

  init {
    "payload" should "be correctly encoded" {
      if (message != null) {
        val encodedBytes = encode<T>(codec, message!!)

        println("encodedBits=${HexUtil.hex(encodedBytes)}")
        encodedBytes.toList() shouldBe payload.toList()
      }
    }

    "payload" should "be correctly decoded" {
      val actual = decodeFully(codec, payload)

      if (message != null) {
        actual shouldBe message
      }
    }

    "payload" should "be correctly decoded and encoded" {
      val obj = decodeFully(codec, payload)
      encode(codec, obj).toList() shouldBe payload.toList()
    }
  }
}


abstract class MultiplePayloadTestSuite<T> : FlatSpec(), Matchers, CodecTestUtil {
  abstract val codec : Codec<T>
  abstract val payloads : Table2<T, ByteArray>

  init {
    "payloads" should "be correctly decoded" {
      forAll(payloads) { message: T, payload: ByteArray ->
        val actual = decodeFully(codec, payload)

        //println("message : ${(message as ByteBuf).toByteArray().toList()}, actual : ${(actual as ByteBuf).toByteArray().toList()},  payload : ${payload.toList()}")

        if (message != null) {
          if (message is ByteArray) {
            // Compare the contents of two ByteArrays.
            // Note that ByteArray.equals check reference equality.
            val decoded = actual as ByteArray
            //println("check : ${decoded.toList()}, ${message.toList()}")
            decoded.toList() shouldBe message.toList()
          } else {
            actual shouldBe message
          }
        }
      }
    }

    "payloads" should "be correctly encoded" {
      forAll(payloads) { message : T, payload : ByteArray ->
        //println("message : ${(message as ByteBuf).toByteArray().toList()}, payload : ${payload.toList()}")

        if (message != null) {
          val encodedBytes = encode<T>(codec, message)

          println("encodedBits=${HexUtil.hex(encodedBytes)}")
          encodedBytes.toList() shouldBe payload.toList()
        }
      }
    }

    "payloads" should "be correctly decoded and encoded" {
      forAll(payloads) { _ : T, payload: ByteArray ->
        val obj = decodeFully(codec, payload)
        encode(codec, obj).toList() shouldBe payload.toList()
      }
    }
  }
}


/**
  * Created by kangmo on 1/29/16.
  */
abstract class EnvelopeTestSuite<T> : PayloadTestSuite<T>() {
  val envelopeCodec = BitcoinMessageEnvelopeCodec /* BUGBUG Need to set the code instance */

  abstract val envelope : BitcoinMessageEnvelope
  abstract val envelopeHeader : ByteArray

  fun envelopeBytes() = envelopeHeader + payload

  init {
    "envelope" should "be correctly encoded" {
//    println(s"encoded111 : ${HexUtil.hex(envelopeCodec.encode(envelope).require.toByteArray)}")
      envelopeCodec.encode(envelope).toList() shouldBe envelopeBytes().toList()
    }

    "envelope" should "be correctly decoded" {
      val actual = envelopeCodec.decode(envelopeBytes())
      actual shouldBe envelope
    }
  }
}
